src/Illuminate/Log/Context/Repository.php

```diff
@@ -0,0 +1,421 @@
+<?php
+
+namespace Illuminate\Log\Context;
+
+use Illuminate\Auth\Events\Authenticated;
+use Illuminate\Auth\Events\Login;
+use Illuminate\Auth\Events\Logout;
+use Illuminate\Contracts\Container\Container;
+use Illuminate\Foundation\Application;
+use Illuminate\Log\Context\Events\Dehydrating;
+use Illuminate\Log\Context\Events\Hydrated;
+use RuntimeException;
+
+class Repository
+{
+    /**
+     * The container instance.
+     *
+     * @var \Illuminate\Foundation\Application
+     */
+    protected $app;
+
+    /**
+     * The contextual data.
+     *
+     * @var array<string, mixed>
+     */
+    protected $data = [];
+
+    /**
+     * The secret contextual data.
+     *
+     * @var array<string, mixed>
+     */
+    protected $secrets = [];
+
+    /**
+     * The auth guard to restore as default.
+     *
+     * @var string|null
+     */
+    protected $restoreGuard = null;
+
+    /**
+     * Create a new Context instance.
+     */
+    public function __construct(Application $app)
+    {
+        $this->app = $app;
+    }
+
+    /**
+     * Set the given key's value.
+     *
+     * @param  string|array<string, mixed>  $key
+     * @param  mixed  $value
+     * @return $this
+     */
+    public function add($key, $value = null)
+    {
+        $values = is_array($key) ? $key : [$key => $value];
+
+        foreach ($values as $key => $value) {
+            $this->data[$key] = $value;
+        }
+
+        return $this;
+    }
+
+    /**
+     * Set the given key's value as a secret.
+     *
+     * @param  string|array<string, mixed>  $key
+     * @param  mixed  $value
+     * @return $this
+     */
+    public function addSecret($key, $value = null)
+    {
+        $values = is_array($key) ? $key : [$key => $value];
+
+        foreach ($values as $key => $value) {
+            $this->secrets[$key] = $value;
+        }
+
+        return $this;
+    }
+
+    /**
+     * Forget the given key's context.
+     *
+     * @param  string|array<int, string>  $key
+     * @return $this
+     */
+    public function forget($key)
+    {
+        foreach ((array) $key as $k) {
+            unset($this->data[$k]);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Forget the given key's secret context.
+     *
+     * @param  string|array<int, string>  $key
+     * @return $this
+     */
+    public function forgetSecret($key)
+    {
+        foreach ((array) $key as $k) {
+            unset($this->secrets[$k]);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Set the given key's value if it does not yet exist.
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     * @return $this
+     */
+    public function addIf($key, $value)
```

Comment:
I think the `addIf` and `addSecretIf` methods should be renamed to something like `addIfKeyIsMissing` and `addSecretIfKeyIsMissing`. Please let me know what you think, but for me, when I read `addIf` I think about adding the value if some condition is true, instead of adding only if the key does not exist.

Reply:
I actually think a better approach would be restrictive by default. Make the plain `add` version throw an exception if the key already exists and provide a `overwrite` or `$force` parameter. Users can explicitly specify that they are okay with rewriting the Context and getting a fatal error if they are not aware that they are overwriting it.

Given the fact this functionality is suppose to cross execution boundaries, it would make more sense (to me) to provide a readonly experience by default and offer a way to overwrite an existing value if the developer explicitly wants to do it.

Reply:
Yeah, that's likely the ok, but that also adds an extra level of error handling that people need to deal with.

I'm torn between "override by default", "only override with a flag", and "throw exception if already set". You ultimately want to minimise the chance of _unexpected_ behaviour, but also ensure that there's a good DX.

Having to wrap everything in either a `try`/`catch` when adding to context or having to check if the key has been set before setting a variable would be bad DX.

I think that maybe the best option is to be able to control default behaviour in a service provider:

```php
// Throw exception if trying to replace context
Context::onDuplicate(Context::THROW_EXCEPTION);
// Replace the existing value
Context::onDuplicate(Context::REPLACE);
// Keep the existing value
Context::onDuplicate(Context::KEEP);

// Custom behaviour which lets you control which value gets stored.
Context::onDuplicate(function ($key, $new, $existing, $private) {
    if ($key === 'auth' && $private) {
        throw new RuntimeException('Cannot change user context');
    }

    return $private ? $existing : $new;
});
```

Reply:
@timacdonald I think the Context class could be macroable. This way I can add this feature if it's included.

Reply:
@samlev my experience differs there because whenever I want to make things restrictive / readonly, the exception is never to be caught, but rather to be a signal. If I write a code that tries to overwrite something and I get an exception, I have two choices: either acknowledge that I'm really overwriting something and make a call with `->add($blah, force: true);` or I rethink what I'm doing because overwriting was not expected.

An exception being thrown isn't always an invitation for try / catch, it's a notification that something you didn't know you were doing is happening.

Reply:
Yeah, I understand that. I just think that like with strict models it's a behaviour that you may want to control on a per app/per environment basis:

```php
Context::onDuplicate(Context::THROW);
if (app()->environment('production')) {
    Context::onDuplicate(Context::KEEP);
}
```

Reply:
I've borrowed the `addIf` method naming from the container, which is used to only add the binding if it is not yet present.

```
App::bindIf('log', fn () => /* ... */);

App::scopedIf('log', fn () => /* ... */);
```

The naming is not a hill I want to die on, but I don't mind `addIf`. If we where to change it I would probably opt for `addWhenMissing` and `addSecretWhenMissing`.

I'll also make it macroable so you can create the strict methods.

Reply:
Never fear, composition (or inheritance) is here.

```php
use Illuminate\Log\Context\Repository;

$this->app->bind(Repository::class, StrictRepository::class);

/**
 * @mixin Repository
 */
class StrictRepository
{
    public function __construct(private Repository $context)
    {
        //
    }

    /**
     * Set the given key's value.
     *
     * @param  string|array<string, mixed>  $key
     * @param  mixed  $value
     * @return $this
     */
    public function add($key, $value = null)
    {
        if ($this->has($key)) {
            throw RuntimeException('Whoops!');
        }

        $this->add(...func_get_args());

        return $this;
    }

    public function __call($name, $arguments)
    {
        return $this->context->{$name}(...$arguments);
    }
}
```